### PR TITLE
Feature/week2

### PR DESCRIPTION
**Pull request by U08410006**
### Todo
- [x] ch4
- [x] ch5

### CODE REVIEW

File: u08410006/Week2/ch4/ch4-10.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+        if sha1.hexdigest() == sha1_hash:
+            return fname  # Hit cache
+    print(f"Downloading {fname} from {url}...")
+    r = requests.get(url, stream=True, verify=True)
+    with open(fname, "wb") as f:
+        f.write(r.content)
+    return fname
+
+
+def download_extract(name, folder=None):
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-10.py:37]: function 名稱要完整的講述他在做啥，i.e., `download_and_extract_zip_file`

File: u08410006/Week2/ch4/ch4-10.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    """Download and extract a zip/tar file."""
+    fname = download(name)
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-10.py:39]: file_name 比較直覺

File: u08410006/Week2/ch4/ch4-10.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+import torch
+import numpy as np
+import pandas as pd
+from torch import nn
+from d2l import torch as d2l
+
+
+def download(name, cache_dir=os.path.join("..", "data")):
+    """Download a file inserted into DATA_HUB, return the local filename."""
+    assert name in DATA_HUB, f"{name} does not exist in {DATA_HUB}."
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-10.py:16]: 如果有用到這種全域變數 要放在這個檔案的最上面，不要放在 main 底下

File: u08410006/Week2/ch4/ch4-10.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+    url, sha1_hash = DATA_HUB[name]
+    os.makedirs(cache_dir, exist_ok=True)
+    fname = os.path.join(cache_dir, url.split("/")[-1])
+    if os.path.exists(fname):
+        sha1 = hashlib.sha1()
+        with open(fname, "rb") as f:
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-10.py:22]: `encoding="utf-8"` 還有不要叫 f 改成更好懂的名字

File: u08410006/Week2/ch4/ch4-10.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+            while True:
+                data = f.read(1048576)
+                if not data:
+                    break
+                sha1.update(data)
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-10.py:27]: 強烈建議這種邏輯可以包成另一個 function 不過這次就算了，因為大部分邏輯不是自己寫的，不過以後盡量不要在 function 裡面疊超過三層邏輯，不然別人會很難看懂

File: u08410006/Week2/ch4/ch4-10.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        fp = tarfile.open(fname, "r")
+    else:
+        assert False, "Only zip/tar files can be extracted."
+    fp.extractall(base_dir)
+    return os.path.join(base_dir, folder) if folder else data_dir
+
+
+def download_all():
+    """Download all files in the DATA_HUB."""
+    for name in DATA_HUB:
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-10.py:54]: 把 data_hub 當作參數傳進來比較靈活

File: u08410006/Week2/ch4/ch4-10.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+    test_labels,
+    num_epochs,
+    learning_rate,
+    weight_decay,
+    batch_size,
+):
+    """
+    Defining the Training Loop
+    """
+    train_ls, test_ls = [], []
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-10.py:89]: 什麼事 ls ?

File: u08410006/Week2/ch4/ch4-10.py [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+    train_iter = d2l.load_array((train_features, train_labels), batch_size)
+    # The Adam optimization algorithm is used here
+    optimizer = torch.optim.Adam(
+        net.parameters(), lr=learning_rate, weight_decay=weight_decay
+    )
+    for _ in range(num_epochs):
+        for X, y in train_iter:
+            optimizer.zero_grad()
+            train_loss = loss(net(X), y)
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-10.py:98]: loss function 哪來的？

File: u08410006/Week2/ch4/ch4-10.py (hunk 7 above)
Comment (jamfly) [reply u08410006/Week2/ch4/ch4-10.py:89]: ls 是 loss 嗎？如果是的話不要縮寫，這樣看的人需要通靈

File: u08410006/Week2/ch4/ch4-10.py [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+            train_loss.backward()
+            optimizer.step()
+        train_ls.append(log_rmse(net, train_features, train_labels))
+        if test_labels is not None:
+            test_ls.append(log_rmse(net, test_features, test_labels))
+    return train_ls, test_ls
+
+
+def get_k_fold_data(k, i, X, y):
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-10.py:107]: 參數名稱太簡陋，沒辦法從名字看出它的功用

File: u08410006/Week2/ch4/ch4-10.py [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+
+    DATA_HUB["kaggle_house_train"] = (
+        DATA_URL + "kaggle_house_pred_train.csv",
+        "585e9cc93e70b39160e7921475f9bcd7d31219ce",
+    )
+
+    DATA_HUB["kaggle_house_test"] = (
+        DATA_URL + "kaggle_house_pred_test.csv",
+        "fa19780a7b011d9b009e8bff8e99922a8ee2eb90",
+    )
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-10.py:214]: 這種全域的要放在最上面

File: u08410006/Week2/ch4/ch4-10.py [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+    train_features = torch.tensor(
+        all_features[:n_train].values, dtype=torch.float32
+    )
+    test_features = torch.tensor(
+        all_features[n_train:].values, dtype=torch.float32
+    )
+    train_labels = torch.tensor(
+        train_data.SalePrice.values.reshape(-1, 1), dtype=torch.float32
+    )
+    loss = nn.MSELoss()
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-10.py:248]: 這邊不用再把 loss 指給 loss 這個變數，這樣會讓人誤會他有計算 loss ，叫 loss 的時候都應該是被計算成值，這樣光看名字就知道他是 functuon 還是數值，就不會搞混

Comment (jamfly) [reply u08410006/Week2/ch4/ch4-10.py:248]: 另外你這個 loss 是不是沒有塞到任何地方，跟上面的 function 的 loss 好像沒有接在一起，你有 run 過這個 code 嗎？

File: u08410006/Week2/ch4/ch4-2.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+def relu(X):
+    """
+    activation ourselves using the maximum function
+    """
+    a = torch.zeros_like(X)
+    return torch.max(X, a)
+
+
+def net(X):
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-2.py:14]: get_get or build_net would be good

File: u08410006/Week2/ch4/ch4-2.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    """
+    reshape and implement our model
+    """
+    X = X.reshape((-1, NUM_INPUTS))
+    H = relu(X @ W1 + b1)  # Here '@' stands for matrix multiplication
+    return H @ W2 + b2
+
+
+if __name__ == "__main__":
+
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-2.py:24]: 不要空格

File: u08410006/Week2/ch4/ch4-2.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    )
+    b1 = nn.Parameter(torch.zeros(NUM_HIDDENS, requires_grad=True))
+    W2 = nn.Parameter(
+        torch.randn(NUM_HIDDENS, NUM_OUTPUTS, requires_grad=True) * 0.01
+    )
+    b2 = nn.Parameter(torch.zeros(NUM_OUTPUTS, requires_grad=True))
+
+    params = [W1, b1, W2, b2]
+
+    loss = nn.CrossEntropyLoss()
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-2.py:41]: 直接把 crossEntropyLoss 塞進去就好，不用賦值

File: u08410006/Week2/ch4/ch4-3.py [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+
+if __name__ == "__main__":
+    net = nn.Sequential(
+        nn.Flatten(), nn.Linear(784, 256), nn.ReLU(), nn.Linear(256, 10)
+    )
+
+    net.apply(init_weights)
+
+    BATCH_SIZE, LR, NUM_EPOCHS = 256, 0.1, 10
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-3.py:21]: 常數要在最上面，你可以放在 main function 的最上面，因爲沒有其他 function 會用到，或是放到檔案的最上面

File: u08410006/Week2/ch4/ch4-4.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
+import math
+
+import torch
+import numpy as np
+from torch import nn
+from d2l import torch as d2l
+
+
+def evaluate_loss(net, data_iter, loss):
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-4.py:9]: loss 改成 loss_function 比較好讀

File: u08410006/Week2/ch4/ch4-4.py [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    return metric[0] / metric[1]
+
+
+def train(
+    train_features, test_features, train_labels, test_labels, num_epochs=400
+):
+    """
+    define training function
+    """
+    loss = nn.MSELoss()
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-4.py:26]: 跟上面一樣

File: u08410006/Week2/ch4/ch4-4.py [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+    trainer = torch.optim.SGD(net.parameters(), lr=0.01)
+    animator = d2l.Animator(
+        xlabel="epoch",
+        ylabel="loss",
+        yscale="log",
+        xlim=[1, num_epochs],
+        ylim=[1e-3, 1e2],
+        legend=["train", "test"],
+    )
+    for epoch in range(num_epochs):
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-4.py:47]: `range(1, num_epochs + 1)` 更直覺

File: u08410006/Week2/ch4/ch4-4.py [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+                (
+                    evaluate_loss(net, train_iter, loss),
+                    evaluate_loss(net, test_iter, loss),
+                ),
+            )
+    print("weight:", net[0].weight.data.numpy())
+
+
+if __name__ == "__main__":
+
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-4.py:61]: 不要空格

File: u08410006/Week2/ch4/ch4-5.py [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+def train(lambd):
+    """
+    Defining the Training Loop
+    """
+
+    def net(X, w, b):
+        """
+        Linear_Regression
+        """
+        return d2l.linreg(X, w, b)
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-5.py:31]: 放到外面，寫在 train 裡面沒有好處反而不靈活

File: u08410006/Week2/ch4/ch4-5.py [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+
+    NUM_EPOCHS, LR = 100, 0.003
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-5.py:33]: 變成參數傳進來比較靈活

File: u08410006/Week2/ch4/ch4-5.py [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+                ),
+            )
+    print("L2 norm of w:", torch.norm(w).item())
+
+
+def train_concise(wd):
+    """
+    specify the weight decay hyperparameter directly through weight_decay when instantiating our optimizer
+    """
+    NUM_EPOCHS, LR = 100, 0.003
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-5.py:64]: 變成參數比較靈活

File: u08410006/Week2/ch4/ch4-6.py [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+            # Add a dropout layer after the second fully connected layer
+            H2 = dropout_layer(H2, dropout2)
+        out = self.lin3(H2)
+        return out
+
+
+if __name__ == "__main__":
+    NUM_INPUTS, NUM_OUTPUTS, NUM_HIDDENS1, NUM_HIDDENS2 = 784, 10, 256, 256
+    NUM_EPOCHS, LR, BATCH_SIZE = 10, 0.5, 256
+    dropout1, dropout2 = 0.2, 0.5
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-6.py:65]: 為啥 dropout 不是常數？

**Review by jamfly — CHANGES_REQUESTED** [review]

File: u08410006/Week2/ch4/ch4-10.py [hunk 24]
```diff
@@ -0,0 +1,276 @@
...
+DATA_HUB["kaggle_house_train"] = (
+    DATA_URL + "kaggle_house_pred_train.csv",
+    "585e9cc93e70b39160e7921475f9bcd7d31219ce",
+)
+
+DATA_HUB["kaggle_house_test"] = (
+    DATA_URL + "kaggle_house_pred_test.csv",
+    "fa19780a7b011d9b009e8bff8e99922a8ee2eb90",
+)
+
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-10.py:25]: 也可以寫成
```
DATA_HUB = {
    "kaggle_house_train": ...
    "kaggle_house_test": ...
}
```

看起來比較乾淨一些

File: u08410006/Week2/ch4/ch4-10.py (hunk 6 above)
Comment (jamfly) [reply u08410006/Week2/ch4/ch4-10.py:67]: 這邊如果是傳進來當參數的話，就不應該是使用常數的命名，如果要直接用常數的話就不需要參數。
這樣也可以 `data_hub=DATA_HUB`

File: u08410006/Week2/ch4/ch4-5.py [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+
+
+def net(X, w, b):
+    """
+    Linear_Regression
+    """
+    return d2l.linreg(X, w, b)
+
+
+def train(lambd, NUM_EPOCHS, LR):
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-5.py:29]: 跟上面一樣

File: u08410006/Week2/ch4/ch4-5.py [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+                epoch + 1,
+                (
+                    d2l.evaluate_loss(net, train_iter, d2l.squared_loss),
+                    d2l.evaluate_loss(net, test_iter, d2l.squared_loss),
+                ),
+            )
+    print("L2 norm of w:", torch.norm(w).item())
+
+
+def train_concise(wd, NUM_EPOCHS, LR):
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-5.py:59]: 跟上面一樣

**Review by jamfly — CHANGES_REQUESTED** [review]

File: u08410006/Week2/ch4/ch4-5.py [hunk 27]
```diff
@@ -0,0 +1,110 @@
+import torch
+from torch import nn
+from d2l import torch as d2l
+
+
+def init_params():
+    """
+    randomly initialize our model parameters
+    """
+    w = torch.normal(0, 1, size=(NUM_INPUTS, 1), requires_grad=True)
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-5.py:10]: `NUM_INPUTS` 可以是參數，會比較靈活。`init_params(num_input: int = NUM_INPUTS)`

File: u08410006/Week2/ch4/ch4-5.py [hunk 28]
```diff
@@ -0,0 +1,110 @@
...
+        xlim=[5, num_epochs],
+        legend=["train", "test"],
+    )
+    for epoch in range(num_epochs):
+        for X, y in train_iter:
+            # The L2 norm penalty term has been added, and broadcasting
+            # makes `l2_penalty(w)` a vector whose length is `BATCH_SIZE`
+            loss = d2l.squared_loss(net(X), y) + lambd * l2_penalty(w)
+            loss.sum().backward()
+            d2l.sgd([w, b], lr, BATCH_SIZE)
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-5.py:47]: batch_szie 也是，變成參數會比較靈活

File: u08410006/Week2/ch4/ch4-5.py [hunk 29]
```diff
@@ -0,0 +1,110 @@
...
+                ),
+            )
+    print("L2 norm of w:", torch.norm(w).item())
+
+
+def train_concise(wd, num_epochs, lr):
+    """
+    specify the weight decay hyperparameter directly through weight_decay when instantiating our optimizer
+    """
+    net = nn.Sequential(nn.Linear(NUM_INPUTS, 1))
```
Comment (jamfly) [inline u08410006/Week2/ch4/ch4-5.py:63]: 跟上面一樣，變成參數比較靈活，或是直接把 net 傳進來

**Review by jamfly — CHANGES_REQUESTED** [review]

**Review by jamfly — APPROVED** [review]
